### ISSUE TITLE
Variance with generic parent class

### ISSUE THREAD

**Issue by actsasbuffoon** [crystal-lang/crystal#1943]
I'm getting an unexpected error from the Crystal compiler. Here's an example:

```crystal
class Parent(T)
end

class Child < Parent(String)
end

def example(value : Parent(String)+)
end

example Child.new
```
This results in a compiler error:

```text
Bug: unsupported restriction: Child vs. Parent(String)+
[4523669323] *raise<String>:NoReturn +139
[4528370945] *Crystal::Type+@Crystal::Type#restrict<Crystal::Type+, Crystal::ASTNode+, Crystal::MatchContext>:NoReturn +529
[4528468570] *Crystal::MatchesLookup::match_arg<Crystal::Type+, Crystal::Arg, Crystal::MatchContext>:(Nil | Crystal::Type+) +1434
[4528465964] *Crystal::MatchesLookup::match_def<Crystal::CallSignature, Crystal::DefWithMetadata, Crystal::MatchContext>:Crystal::Match? +1100
[4529146351] *Crystal::Program@Crystal::MatchesLookup#lookup_matches_without_parents<Crystal::Program, Crystal::CallSignature, Crystal::Program, Crystal::Program, Nil>:Crystal::Matches +335
[4529143777] *Crystal::Program@Crystal::MatchesLookup#lookup_matches<Crystal::Program, Crystal::CallSignature, Crystal::Program, Crystal::Program, Nil>:Crystal::Matches +81
[4529239220] *Crystal::Call#lookup_matches_with_signature<Crystal::Call, Crystal::Program, Crystal::CallSignature, Bool>:Crystal::Matches +340
[4528419330] *Crystal::Call#lookup_matches_in_type<Crystal::Call, Crystal::Program, Array(Crystal::Type+), Nil, String, Bool>:Array(Crystal::Def+) +642
[4523957767] *Crystal::Call#recalculate<Crystal::Call>:Array(Crystal::Def+)? +2679
[4526783554] *Crystal::TypeVisitor#visit<Crystal::TypeVisitor, Crystal::Call>:Bool +6018
[4525078922] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::TypeVisitor>:Nil +2266
[4525168776] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::TypeVisitor>:Nil +92120
[4526662915] *Crystal::Program#infer_type<Crystal::Program, Crystal::Expressions>:Crystal::Expressions +179
[4526085987] *Crystal::Compiler#compile<Crystal::Compiler, Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result +2979
[4530222929] *Crystal::Command#run_command<Crystal::Command>:Nil +145
[4530052947] *Crystal::Command#run<Crystal::Command>:(Nil | Bool | Process::Status | Array(String) | IO::FileDescriptor+ | Crystal::Compiler::Result | Array(Crystal::ImplementationTrace) | Array(Crystal::Init::View+:Class)) +8563
[4530043276] main +17164

Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/manastech/crystal/issues
```

The code in my example seems weird, but that's because I've simplified it to get to the essence of the problem. I'm toying around with writing a parser combinator library, and this functionality would be really useful.

If `Child` is a subclass of `Parent(String)` then it seems like it should work with a `Parent(String)+` type restriction.

**Comment by kirbyfan64** [comment]
IIRC, @asterite said the + was supposed to be an internal thing; you should never have to use it explicitly.

**Comment by kirbyfan64** [comment]
Found it: https://groups.google.com/forum/m/#!topic/crystal-lang/mSrYYcyIV7c.

**Comment by actsasbuffoon** [comment]
@kirbyfan64 Good to know, thank you.

I may have over-simplified my example code. Something a little closer shows why I thought including the `+` might help.

```crystal
class Parent(T); end

class Child < Parent(String); end

class Example
  @value :: Parent(String)

  def initialize(@value)
  end
end

Example.new Child.new
```

This results in:

```text
Error in ./failure.cr:12: instantiating 'Example:Class#new(Child)'

Example.new Child.new
        ^~~

instantiating 'Example#initialize(Child)'

in ./failure.cr:8: type must be Parent(String), not (Parent(String) | Child)

  def initialize(@value)
                 ^

================================================================================

Child trace:

  ./failure.cr:8

      def initialize(@value)
                     ^~~~~

  ./failure.cr:8

      def initialize(@value)
```

This seems to be saying that `@value` must be a `Parent(String)`, and `Child` doesn't fulfill that requirement.

**Comment by asterite** [comment]
Yes, inheriting from a generic class doesn't work pretty well :-(

**Comment by jhass** [comment]
The first example works now, the second example gives a compile time error when ported to new syntax. When adding a cast it gives a compile time exception instead of a proper error message: https://carc.in/#/r/zym

**Comment by asterite** [comment]
Closed in favor of #2665
